flash_attention integration

drisspg opened the pull request:
# Summary:
- I added a new submodule Cutlass pointing to 2.10 release. The inclusion of flash_attention code should be gated by the flag: USE_FLASH_ATTENTION. This is defaulted to off resulting in flash to not be build anywhere. This is done on purpose since we don't have A100 machines to compile and test on.

- Only looked at CMake did not attempt bazel or buck yet.

-  I included the mha_fwd from flash_attention that has ben refactored to use cutlass 2.10. There is currently no backwards kernel on this branch. That would be a good follow up.


cc @ngimel

drisspg commented on CMakeLists.txt:
```diff
@@ -782,6 +782,9 @@ if(USE_COREML_DELEGATE)
 
+# # Lets try adding cutlass manually
+# string(APPEND CMAKE_CXX_FLAGS " -DUSE_CUTLASS")
+
```
This was cannon fodder.. but something like this flag for `-DUSE_FLASH_ATTENTION` would probably be useful if it was wired up correctly

drisspg commented on aten/src/ATen/CMakeLists.txt:
```diff
@@ -130,8 +130,8 @@ file(GLOB native_cuda_h "native/cuda/*.h" "native/cuda/*.cuh")
-file(GLOB native_nested_cuda_cpp "native/nested/cuda/*.cpp")
+file(GLOB native_nested_cuda_cu "native/nested/cuda/*.cu" "native/nested/flash_attn/*.cu")
+file(GLOB native_nested_cuda_cpp "native/nested/cuda/*.cpp" "native/nested/flash_attn/*.cpp")
 file(GLOB native_sparse_cuda_cu "native/sparse/cuda/*.cu")
```
This took too long to find, but glad I know why arbitrarily adding folders doesn't work in native

drisspg commented on aten/src/ATen/CMakeLists.txt:
```diff
@@ -415,6 +415,12 @@ if(NOT MSVC AND NOT EMSCRIPTEN AND NOT INTERN_BUILD_MOBILE)
+  # add_subdirectory("${CMAKE_CURRENT_SOURCE_DIR}/../../../third_party/cutlass" ${CMAKE_BINARY_DIR}/cutlass)
+  list(APPEND ATen_THIRD_PARTY_INCLUDE ${TORCH_ROOT}/third_party/cutlass/include)
+  list(APPEND ATen_CUDA_INCLUDE ${CMAKE_CURRENT_SOURCE_DIR}/../../../third_party/cutlass/include)
+  message("!!!!!!!!!!!!!!!!!THIRD PARTY INCLUDES BRO: ${ATen_THIRD_PARTY_INCLUDE}!!!!!!!!!!!!!!!!!")
```
whoops.. cmake was getting frustrating.. I think that line is actually what is needed so that when building the cu kernel cutlasss includes can be found

cpuhrsch commented on aten/src/ATen/CMakeLists.txt:
```diff
@@ -112,7 +112,7 @@ file(GLOB native_sparse_cpp "native/sparse/*.cpp")
             "native/quantized/*.cpp"
             "native/quantized/cpu/*.cpp")
-file(GLOB native_nested_cpp "native/nested/*.cpp")
+file(GLOB native_nested_cpp "native/nested/*.cpp" "native/nested/flash_attn/*.cpp")
```
I suspect this is causing build issues, because some of the .cpp files in that folder pull in CUDA headers (e.g. `#include <ATen/cuda/CUDAContext.h>` in fmha_api.cpp) and we separate our CPU and CUDA builds. One way around this is having `flash_scaled_dot_product_self_attention` only registered under the CUDA and NestedTensorCUDA flags and generally hove the whole thing under `native/nested/cuda/flash_attn`.

drisspg replied:
updated

cpuhrsch commented on aten/src/ATen/native/nested/cuda/flash_attn/fmha_api.cpp:
```diff
@@ -0,0 +1,262 @@
+
+
+namespace at {
+namespace native {
```
nit: I suggest we put this into its own namespace just to avoid collisions.

cpuhrsch commented:
You can also guard on CUDA 11.3 (which also seems to fail) by using the [CUDA_VERSION](https://github.com/pytorch/pytorch/blob/085e2f7bddc45f859fcdb786926d60d709b2daa0/c10/cuda/CUDAGraphsC10Utils.h#L20) macro.

drisspg commented:
> You can also guard on CUDA 11.3 (which also seems to fail) by using the [CUDA_VERSION](https://github.com/pytorch/pytorch/blob/085e2f7bddc45f859fcdb786926d60d709b2daa0/c10/cuda/CUDAGraphsC10Utils.h#L20) macro.

Where do you think I should guard the build in, here:https://github.com/drisspg/pytorch/blob/2d30cde0374b39809efebb8a0e588c03fbcbb347/aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cpp#L218

and just 
``` c++
#if defined(CUDA_VERSION) && CUDA_VERSION >= 11400
    return fmha::mha_fwd(...)
#endif

return std::vector<Tensor> {};
```

or should this be done in _flash_scaled_dot_prod

cpuhrsch commented:
@drisspg - I think you could guard on the CUDA version within the cmake files and just entirely disable the build of flash_attn if at least CUDA 11.6 isn't available. Then within `_flash_scaled_dot_prod` you like you detailed using `#if defined(CUDA_VERSION) && CUDA_VERSION >= 11400` (I think it'll need to be 11600).

drisspg commented:
I think there is some circular dependency issue going on with the header files. I had to include extra header files to compile locally. My included headers deviate from flash attn. I also I initially copied these files before the bfloat16 inclusion.

drisspg commented on aten/src/ATen/CMakeLists.txt:
```diff
@@ -415,6 +424,11 @@ if(NOT MSVC AND NOT EMSCRIPTEN AND NOT INTERN_BUILD_MOBILE)
+  # add_subdirectory("${CMAKE_CURRENT_SOURCE_DIR}/../../../third_party/cutlass" ${CMAKE_BINARY_DIR}/cutlass)
+  # list(APPEND ATen_THIRD_PARTY_INCLUDE ${TORCH_ROOT}/third_party/cutlass/include)
+  if(CMAKE_CUDA_COMPILER_VERSION VERSION_GREATER_EQUAL 11.6)
+    list(APPEND ATen_CUDA_INCLUDE ${CMAKE_CURRENT_SOURCE_DIR}/../../../third_party/cutlass/include)
```
This condition should match flash_attn condition

malfet replied:
Correct me if I'm wrong, but this would make cutlass a requirement for all users of libtorch. It's better to cmake this a private include dependency of `libtorch_cuda`

drisspg replied:
@malfet Internal seems to be appeased by the ifdef changes, otherwise a code pointer for this change would be great and I think its ready to land

drisspg commented on aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cpp:
```diff
@@ -202,5 +205,36 @@ Tensor NestedTensor_to_padded_tensor_cuda(
   return NestedTensor_to_padded_tensor_generic(t, padding, output_size);
 }
+
+Tensor flash_scaled_dot_product_self_attention(
```
This func should take q,k v and helper should split it qkv

drisspg commented on aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cpp:
```diff
@@ -204,5 +207,36 @@ Tensor NestedTensor_to_padded_tensor_cuda(
+    const int64_t max_seqlen_batch,
+    double dropout_p,
+    bool causal) {
+#if defined(CUDA_VERSION) && CUDA_VERSION >= 11060 && !defined(_MSC_VER) && \
```
Not totally sure what to do here. I have a cmake flag that looks for similar conditions and sets the USE_FLASH_ATTENTION flag to true and if thats the case we will build all the flash_atten sources. However if you were on an machine that has linux and is using cuda 11.6 and is linux then this if_define be true but someone could set USE_FLASH_ATTENTION to FALSE and if thats the case this will result in undefined_symbol for mha_fwd.

Not to sure how to resolve this yet

cpuhrsch replied:
Maybe: Use `USE_FLASH_ATTENTION` here as well. If that flag isn't True, then we're not building with it.

drisspg replied:
ASAI I don't think that cmake Options are available to compilier for macro expansion but I am trying to figure out if there is a different way to set USE_FLASH_ATTENTION so that `#if defined(USE_FLASH_ATTENTION)` is true

drisspg replied:
figured it out nvm

cpuhrsch commented on aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cpp:
```diff
@@ -204,5 +207,36 @@ Tensor NestedTensor_to_padded_tensor_cuda(
+      c10::nullopt);
+  return output[0];
+#endif
+  return Tensor{};
```
I'd throw an exception here in case it gets called by accident.

drisspg commented on aten/src/ATen/native/nested/NestedTensorTransformerFunctions.cpp:
```diff
@@ -260,5 +261,128 @@ Tensor NestedTensor_to_mask(const Tensor& nt, c10::optional<int64_t> mask_dim, c
+    atten_size.index({at::indexing::Slice(), -1}) = div3;
+
+    auto qkv_buffer_reshaped =
+        get_buffer(query).view({Nnz_q, 3, num_heads, head_dim});
```
@jbschlosser 
The option to avoid using num_heads is 
` auto qkv_buffer_reshaped = get_buffer(query).view({Nnz_q, 3, 1, -1});`

The problem with this is that currently flash_attention has the restriction that:
`    TORCH_CHECK(head_size == 16 || head_size == 32 || head_size == 64 || head_size == 128, "Currently supported head_sizes: {16, 32, 64, 128}, passed in query head size of ", head_size);
`
So by doing this and avoiding num heads we just expand out the head_dim. Unfortunately for most reasonably sized q,k,v this expands the head dim passed 128.

I might be missing something: https://github.com/fairinternal/xformers/blob/cd8e3d218acc330029be5340fe6315df444b4bbf/xformers/ops.py#L339
This is what I was using as a reference.

jbschlosser replied:
Sorry, I don't quite understand. I'd expect this for MHA (focusing on query but analogous for key / value):
1. Receive query of shape (N, L, E); N = original batch size, L = seq len, E = full embedding dim
2. Do query projection -> shape (N, L, E'), with E' = post-projection embedding dim
3. Reshape query to split heads -> shape (N, L, n_heads, head_dim); E' = n_heads * head_dim
4. Transpose query -> shape (N, n_heads, L, head_dim)
5. Combine query leading dims -> shape (N * n_heads, L, head_dim)
6. Pass query to SDP
7. Do reshaping + post-projection

So SDP should receive a query input of shape (N * n_heads, L, head_dim). Flash attention requires an input of (Nnz, n_heads, head_dim). IIUC xFormers sets n_heads=1 for passing to flash attention. So flash attention gets input of shape (N * n_heads * L, 1, head_dim). This should work fine with the head_dim constraints that flash attention has AFAICT.

drisspg replied:
I changed assumptions on input which was a comment christian made.  This is an flash MHA nn module that does packed projection:
``` Python3
class FlashMHA(torch.nn.Module):
    def __init__(self, num_heads, qkv, proj):
        super().__init__()
        self.qkv = qkv
        self.proj = proj
        self.num_heads = num_heads

    def forward(self, query: torch.Tensor, key: torch.Tensor, value: torch.Tensor):
        assert query is key
        assert key is value

        query_projected = torch.nn.functional.linear(query, self.qkv.weight, self.qkv.bias)
        batch_size = query_projected.size(0)
        embed_dim = query_projected.size(-1)
        head_dim = embed_dim//(3*num_heads)
        
        query_projected = torch.reshape(query_projected, (batch_size, -1, 3, num_heads, head_dim))
        attn, _ = torch._scaled_dot_product_attention(query_projected, query_projected, query_projected)
        attn = attn.reshape((batch_size, -1, num_heads*head_dim))
        attn = torch.nn.functional.linear(attn, self.proj.weight, self.proj.bias)
        return attn, None
```

```
        #query_projected -> batch x (seq_len_q) x 3*num_heads*head_dim
        #query_projected -> batch x (seq_len_q) x 3 x num_heads x head_dim
        #attn -> batch x (seq_len_q) x 1 x num_heads x head_dim
        #attn -> batch x (seq_len_q) x num_heads*head_dim
  ```

drisspg commented on aten/src/ATen/native/transformers/attention.cpp:
```diff
@@ -686,8 +687,13 @@ std::tuple<Tensor, Tensor, Tensor, Tensor> native_decoder_only_multi_head_attent
+        const Tensor& query_, const Tensor& key, const Tensor& value, int64_t num_heads,
         const c10::optional<Tensor>& attn_mask_, double dropout_p, bool need_attn_weights, bool is_causal) {
+    // [TODO] Actually replace with the SDP_Backend logic, Figure out how to not use num_heads
+    if(true){
```
this will obviously be changed, I spent some time trying to get around have to use num_heads but I couldn't figure out any way to extract out the information from the nested tensors. This is gross and is bleeding in MHA logic to scaled_dot_prod_attention. idk, maybe make this optional and be fine with its grossness or someone smarter than me figure it out. 

Of course this is all rides on the assumption that an input stacked q,k,v is of shape: `(batch_size x ragged_seq_len x (potentially 3* n_heads * head_dim))` Maybe this is where my flaw lies

cpuhrsch replied:
You can require a higher dimensional input that's already reshaped. 

Notice how q, k and v are [already reshaped](https://github.com/pytorch/pytorch/blob/master/torch/nn/functional.py#L5126) before passed into [_scaled_dot_product](https://github.com/pytorch/pytorch/blob/master/torch/nn/functional.py#L5185). The `num_heads` check then because a check of `q.size(-1)`.

drisspg replied:
Yeah I am curious if someone tries to shape their tensors so that it will pass all the heuristics required for running the flash_attn kernel but `USE_FLASH_ATTENTION=0` or something how hard does this make it to fallback to the composite implicit version or how much shape mangling are we gunna need for the composite implicit version

cpuhrsch commented on aten/src/ATen/native/nested/NestedTensorTransformerFunctions.cpp:
```diff
@@ -260,5 +261,128 @@ Tensor NestedTensor_to_mask(const Tensor& nt, c10::optional<int64_t> mask_dim, c
+    // If we are passing in query, key, value all the same tensors than we have
+    // packed them into one tensor and need to slice for flash attention
+    Tensor atten_buffer = at::_flash_scaled_dot_product_attention(
+        qkv_buffer_reshaped.index({at::indexing::Slice(), 0}),
```
Require users to do the slicing beforehand. I don't see `_flash_scaled_dot_product_attention` requiring that they are all the same. Do you even need this helper or can you call into `_flash_scaled_dot_product_attention` and then do the `cumulative_sequence_length_q` and `max_seqlen_batch_q` within it?

drisspg replied:
I was getting a `wrap_buffer` undefined in the libcuda.so and for my sanity and just to help implement I broke these up. But since this is only a cuda kernel it should probably all live close to one place. Yeah slicing is probably the way to go. Currently slicing can't be done on nested tensors. I think some of Yifens reshape work will help here but that is stalled up by some internal error.

erichan1 commented on aten/src/ATen/native/nested/NestedTensorTransformerFunctions.cpp:
```diff
@@ -260,5 +261,128 @@ Tensor NestedTensor_to_mask(const Tensor& nt, c10::optional<int64_t> mask_dim, c
+        causal);
+  // Output of flash_attention is a regular tensor lets wrap it back up to
+  // form a nested tensor, the size of which should match the value tensor
+  return wrap_buffer(atten_buffer.view(-1), get_nested_size_tensor(value));
```
Is this right? Fairly sure size should match the query tensor.

drisspg replied:
just updated

drisspg commented on aten/src/ATen/native/transformers/sdp_utils.h:
```diff
@@ -0,0 +1,62 @@
+
+
+bool use_flash_attention(sdp_params params) {
+  static const std::unordered_set<int64_t> flash_embed_dim_sizes{16, 32, 64, 128};
```
try constexpr

drisspg commented:
Next up is add the non nested code, should be pretty straightforward

guangzlu commented:
Hi Driss, I am interested in using the transformer with flash attention, could it be built and test now?

malfet commented on aten/src/ATen/CMakeLists.txt:
```diff
@@ -156,6 +154,15 @@ file(GLOB native_transformers_hip_cpp "native/transformers/hip/*.cpp")
 file(GLOB native_quantized_cudnn_hip_cpp "native/quantized/cudnn/hip/*.cpp")
 file(GLOB native_utils_cpp "native/utils/*.cpp")
 
+# Add flash attention sources if condition
+if(USE_FLASH_ATTENTION)
+  file(GLOB native_transformers_cuda_cu "native/transformers/cuda/*.cu" "native/transformers/cuda/flash_attn/*.cu")
+  file(GLOB native_transformers_cuda_cpp "native/transformers/cuda/*.cpp" "native/transformers/cuda/flash_attn/*.cpp")
+else()
+  file(GLOB native_transformers_cuda_cu "native/transformers/cuda/*.cu")
+  file(GLOB native_transformers_cuda_cpp "native/transformers/cuda/*.cpp")
+endif()
```
To follow the same pattern, just create a new glob var and then append it conditionally to the list

drisspg commented on CMakeLists.txt:
```diff
@@ -707,6 +707,14 @@ set(BUILD_ONEDNN_GRAPH OFF)
+# Moved this cmake set option down here because CMAKE_CUDA_COMPILER_VERSION is not avaialble until now
+cmake_dependent_option(USE_FLASH_ATTENTION "Whether to build the flash_attention kernel for scaled dot product attention" ON
+"USE_CUDA;CMAKE_CUDA_COMPILER_VERSION VERSION_GREATER_EQUAL 11.6;NOT USE_ROCM;LINUX" OFF)
+IF(USE_FLASH_ATTENTION)
+    ADD_DEFINITIONS(-DUSE_FLASH_ATTENTION)
+ENDIF()
```
Yeah it should be an option that is dependent on whether or not you are building cuda

drisspg replied:
I was envisioning this being something that someone building from source would like to be able to turn on or off even if all the criteria are met to build

drisspg committed:
[WIP] flash_attention integration (#81434)

Summary:
This is my initial attempt at integrating the flash attention mha_fwd kernel into torch core.

## Notable things:
- I added a new submodule Cutlass. I spent alot of time trying to decipher pytorch's cmake setup. I am sure I did not do this ideally. For instance there will probably be a new flagl -DUSE_FLASH that can be enabled , etc..

- I made a number of changes to dtypes to make the out of tree code work with intree. I will use this PR gui to double check my work.

- Only looked at CMake did not attempt bazel or buck yet.

-  I only included the mha_fwd from flash_attention for sake of minimizing scope. However since a majority of the header files are used for the backward kernel as well as the block kernel if there were follow up PRs they would be smaller

- Next evolution of: https://github.com/pytorch/pytorch/pull/81053 and ran this on the same micro benchmark on compute cluster with A100 and saw similar results.

I wan't CI to run on this but and not sure if I mark as draft if it will run

Pull Request resolved: https://github.com/pytorch/pytorch/pull/81434

Reviewed By: cpuhrsch

Differential Revision: D38253964

Pulled By: drisspg

fbshipit-source-id: 5461a7da3f93c03d249edf66ac6c2ad944abb447

Jack47 commented:
@drisspg Hi，I'm interested to see this PR, FlashAttention is awesome and inspiring, it would be great to let more users benefit it's speedup and memory reduce!  

May I know the progress of this PR? I would like to contribute if helps needed

vadimcn commented:
@cpuhrsch:  Can you please clarify what features SDP currently supports in the nightly?  Can it be used for training?  Does it support key padding masks?

cpuhrsch commented:
@vadimcn - It doesn't support training. It doesn't support key padding masks, but you can pass it an attention mask.

vadimcn commented:
@cpuhrsch Any idea when these features are coming?   I'm trying to decide between basing my code directly on FlashAttention, with all its limitations, or holding for a bit longer till PyTorch implements training and efficient key masking.

cpuhrsch commented:
@vadimcn - We just released 1.13 and scaled dot product attention is not part of it. It'll be part of the next release.

vadimcn commented:
>  We just released 1.13 and scaled dot product attention is not part of it. It'll be part of the next release.

I'm fine with using nightly for now.

lucidrains commented:
@cpuhrsch for the backwards pass, you are still dispatching to Tri Dao's original flash attention implementation if the right hardware presents? is this because the cutlass backwards pass ended up being unacceptably slow?

cpuhrsch commented:
@lucidrains - We're currently working on supporting the backwards pass and are looking into this. We aim to use the fastest implementation eventually.

lucidrains commented:
@cpuhrsch ok, sounds good re: backwards pass

are you falling back to Tri's implementation or no?

cpuhrsch commented:
@lucidrains - For the backwards we're right now we're getting started with xFormer's (memory efficient attention) implementation simply because it's already landed in the repository.

lucidrains commented:
@cpuhrsch ok, hope you get it sorted out

this may be the most important feature you ever add to pytorch, i guarantee it

chrjxj commented:
> @lucidrains - For the backwards we're right now we're getting started with xFormer's (memory efficient attention) implementation simply because it's already landed in the repository.

forks, do we have documents/guides on how to enable/use FA in training & inference?

SimJeg commented:
@cpuhrsch I failed to use flash attention with pytorch 2.0.0 (CUDA 11.7, cudnn 8.5, GPU A10g on AWS g5.2xlarge instance). Here are the code and error I get : 

- Code 

```
import torch
from time import time

dim = 256
num_heads = 8
n_warmup = 10
n_repeats = 1000

attention = torch.nn.MultiheadAttention(dim, num_heads).cuda().eval()
inputs = torch.randn(dim, dim, dim).cuda()

for use_flash in [False, True]:
    with torch.inference_mode(), torch.backends.cuda.sdp_kernel(use_flash, use_flash):
        for _ in range(n_warmup):
            outputs = attention(inputs, inputs, inputs)
        torch.cuda.synchronize()

        start = time()
        for _ in range(n_repeats):
            outputs = attention(inputs, inputs, inputs)
        torch.cuda.synchronize()
        print(f'Forward pass in ms: {(time() - start) * 1000 / n_repeats:.2f}ms')
```

- Error 

```
-> 5200 attn_output, attn_output_weights = _scaled_dot_product_attention(
...
   5201     q, k, v, attn_mask, dropout_p, need_weights, is_causal)
   5202 attn_output = attn_output.permute(2, 0, 1, 3).contiguous().view(bsz * tgt_len, embed_dim)
   5204 attn_output = linear(attn_output, out_proj_weight, out_proj_bias)

RuntimeError: No available kernel.  Aborting execution.
```

What should I do ? Using pytorch 1.13.1 it worked but did not produce any runtime improvement,
Thanks for your help,
Simon

SimJeg commented:
@drisspg  thanks for your help.

I do not have the kernel error anymore with pytorch 2.0 (no idea why...).  Here are some results from my previous script using `need_weights=False` in the forward pass (I did not get any warning) : 

| enable_mem_efficient | enable_flash | runtime (ms) |
|----------------------|--------------|--------------|
| False                | False        | 9.5          |
| False                | True         | 9.5          |
| True                 | False        | 6.4          |
| True                 | True         | 6.4          |

`enable_mem_efficient` is super efficient but `enable_flash` has no effect.

- what exactly is `enable_mem_efficient` ? 
- why `enable_flash` has no effect ?

drisspg commented:
@SimJeg 
This will be documented for the 2.0 release but currently [torch.nn.MultiheadAttention](https://github.com/pytorch/pytorch/blob/f21997099087c6db85e119cd5ef97e2d8a134a85/torch/nn/functional.py#L5200) will under the hood call [torch.nn.functional._scaled_dot_product_attention](https://github.com/pytorch/pytorch/blob/f21997099087c6db85e119cd5ef97e2d8a134a85/torch/nn/functional.py#L4835). 
SDPA is currently private but will be released as a public function for NT 2.0. 

SDPA chooses between three different backends to run:
[Flash-attention](https://arxiv.org/abs/2205.14135)
mem-efficient-attention ( this kernel was written by the xformers team)
math (naive pytorch implementation involving multiple kernel calls)

SDPA should transparently pic the best kernel for you to use. The sdp_kernel context manager is used if you want to specify which kernel to be run and override the default choice. Why would one do this? One wants to experiment with their particular inputs and determine what works best. It is helpful for testing. 

Sooooo I think your table needs another column `enable_math` to really get the full picture of what kernel is being used.

lucidrains commented:
how is mem efficient attention different than flash attention? could you cite the paper used? thanks

drisspg commented:
> how is mem efficient attention different than flash attention? could you cite the paper used? thanks

@lucidrains I would say there is not a fundamental difference between the two algorithms and they are mostly implementation trade offs. One of the differences is in iteration order of the GEMMs, flash-attention being over queries while mem-efficient attention being over the keys. 

https://github.com/facebookresearch/xformers#results, AFAIK there isn't a paper to cite but is developed in xformers.

SimJeg commented:
Hello,

With `enable_math` set to False I get the kernel error.
With pytorch 1.13 I get 9.3ms, which is much higher than 6.4ms with `mem_efficient`.

Thanks for your help again,
Simon

mikekgfb commented:
> @lucidrains this may be the most important feature you ever add to pytorch, i guarantee it

We're super excited about this, now that we have both xFormer Mem Efficient and Flash Attention forward and backward.  The next step is driving adoption to get the maximum value for the community from this function.  

Since you have many key models and example implementations, would you be willing to update those to use F.sdpa() after the PT 2.0 launch?  (Understood that you may still want to support the prior version of PT -- If the symbol isn't defined, we can always go for a fall-back implementation of sdpa using other operators?)

```
try: 
  from torch.functional import scaled_dot_product_attention 
except
  def scaled_dot_product_attention(...) -> torch.Tensor:
    ...
```

mikekgfb commented:
> @SimJeg Table showing 9.5ms and 6.4ms runtimes for combinations of mem_efficient and flash = [ True, False ]

In your earlier table, showing runtimes of 9.5ms and 6.4ms, you might only be exercising the sdpa_math and memory_efficient kernels.  The flash attention kernel is only available on SM80+ (e.g., A100) with FP16 and BF16 data types.  Otherwise, the kernel falls back to the default sdpa_math() catch-all implementation.

lucidrains commented:
@mikekgfb hi Michael

yes indeed, i am excited for this advance in attention as well; thank you for you and your team's hard work on it. i will be adding flash attention to whichever repositories are applicable ([language modeling](https://github.com/lucidrains/palm-rlhf-pytorch), [audio synthesis](http://github.com/lucidrains/audiolm-pytorch), etc). there are also other greenfield projects that flash attention will directly benefit later this year (long context genomics training)

LvanderGoten commented:
I tried to use `attn_mask` (where the the dtype is float, necessary for absolute positional embeddings) with either flash or memory-efficient attention but no avail. Is it currently supported? At least the xformers implementation seems to support it.

cpuhrsch commented:
@LvanderGoten - we'll upstream newest changes from xformers eventually (ideally in time for 2.1).

LvanderGoten commented:
Ok, great! On another note, it would be very helpful if the documentation could include a table of which parameters are supported by the kernels. So far you need to do a trial-and-error approach. Such a table could include type(flash/mem-efficient/math), head size, CUDA compute capability, support for `attn_mask` etc.

drisspg commented:
> > I tried to use `attn_mask` (where the the dtype is float, necessary for absolute positional embeddings) with either flash or memory-efficient attention but no avail. Is it currently supported? At least the xformers implementation seems to support it.
> 
> on this topic, if attention biases requiring gradients are passed through `attn_mask`, will it receive the gradients?

The only path that currenlty supports arbitrary attention masks is the "math" path and yes this one should flow upward grads back to the attn mask. If not please file an issue

mikekgfb commented:
> Ok, great! On another note, it would be very helpful if the documentation could include a table of which options are supported by the kernels. So far you need to do a trial-and-error approach. Such a table could include type(flash/mem-efficient/math), head size, CUDA compute capability, support for `attn_mask` etc.

We are in the process of putting a blog out, as general guidance - but in particular, the operation supported is sdpa() and it picks the implementation.  The kernels chosen will change over time (prospectively picking more aggressive implementations when possible and as/if we relax constraints.)

layumi commented:
Hi @mikekgfb 
Thanks a lot. Could I know the blog address?

mikekgfb commented:
> Hi @mikekgfb Thanks a lot. Could I know the blog address?

Tutorial:
https://pytorch.org/tutorials/intermediate/scaled_dot_product_attention_tutorial.html?highlight=sdpa

Blogs:
https://pytorch.org/blog/accelerating-large-language-models/
https://pytorch.org/blog/accelerated-generative-diffusion-models/
https://pytorch.org/blog/out-of-the-box-acceleration/
https://pytorch.org/blog/accelerated-cpu-inference/
https://pytorch.org/blog/accelerating-generative-ai-2/

(and many more at PyTorch.org under blogs)
